app/PaymentDrivers/CheckoutCom/Utilities.php

```diff
@@ -39,4 +46,129 @@ public function convertToCheckoutAmount($amount, $currency)
+    {
+        if ($this->checkout->payment_hash->data->store_card) {
+            $this->storePaymentMethod($_payment);
+        }
+
+        $data = [
+            'payment_method' => $_payment->source['id'],
+            'payment_type' => PaymentType::parseCardType(strtolower($_payment->source['scheme'])),
+            'amount' => $this->checkout->payment_hash->data->raw_value,
+        ];
+
+        $payment = $this->checkout->createPayment($data, \App\Models\Payment::STATUS_COMPLETED);
+
+        SystemLogger::dispatch(
+            ['response' => $_payment, 'data' => $data],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
+        return redirect()->route('client.payments.show', ['payment' => $this->checkout->encodePrimaryKey($payment->id)]);
+    }
+
+    public function processUnsuccessfulPayment(\Checkout\Models\Payments\Payment $_payment)
+    {
+        PaymentFailureMailer::dispatch(
+            $this->checkout->client,
+            $_payment,
+            $this->checkout->client->company,
+            $this->checkout->payment_hash->data->value
+        );
+
+        $message = [
+            'server_response' => $_payment,
+            'data' => $this->checkout->payment_hash->data,
+        ];
+
+        SystemLogger::dispatch(
+            $message,
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_FAILURE,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
+        throw new PaymentFailed($_payment->status, $_payment->http_code);
+    }
+
+    private function processPendingPayment(\Checkout\Models\Payments\Payment $_payment)
+    {
+        $data = [
+            'payment_method' => $_payment->source['id'],
+            'payment_type' => PaymentType::parseCardType(strtolower($_payment->source['scheme'])),
+            'amount' => $this->checkout->payment_hash->data->value,
+        ];
+
+        $payment = $this->checkout->createPayment($data, \App\Models\Payment::STATUS_PENDING);
+
+        SystemLogger::dispatch(
+            ['response' => $_payment, 'data' => $data],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
+        try {
+            return redirect($_payment->_links['redirect']['href']);
+        } catch (\Exception $e) {
+            return $this->processInternallyFailedPayment($e);
+        }
+    }
+
+    private function processInternallyFailedPayment($e)
+    {
+        if ($e instanceof \Checkout\Library\Exceptions\CheckoutHttpException) {
+            $error = $e->getBody();
+        }
+
+        if ($e instanceof \Exception) {
+            $error = $e->getMessage();
+        }
+
+        PaymentFailureMailer::dispatch(
```

Comment:
lets move this to base driver also.

Reply:
Solved here: https://github.com/invoiceninja/invoiceninja/pull/4202/commits/223ae4cb5bf1823885cf744ce84123f0e654cac2#diff-9229767762db39ad236f5c2a30a13fa853f7f72e4eb1a072ae448ea7d4067493R293-R319